### PR TITLE
Fix space issue

### LINKED ISSUE #36: S3 backend CopyToFile fails when source contains %
Below will fail:
```
        //create file with %
	file, err := remoteLocation.NewFile("orig%20file.txt")
	if err != nil {...	}
	file.Write([]byte("this\n"))
	file.Close()

	//setup target file
	file2, err := remoteLocation.NewFile("new_file.txt")
	if err != nil {...	}

	err = file.CopyToFile(file2)
	if err != nil {
		// WILL FAIL "No such key found"
	}
```
This is because because CopyToFile ultimately calls s3.CopyObject.  s3.CopyObjectInput's CopySource needs to be url encoded.  Oddly, the target does not. https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#CopyObjectInput

```
    // The name of the source bucket and key name of the source object, separated
    // by a slash (/). Must be URL-encoded.
    //
    // CopySource is a required field
    CopySource *string `location:"header" locationName:"x-amz-copy-source" type:"string" required:"true"`
```

We should url-encode CopySource in s3 backend's copyFile() function when calling CopyObjectInput.SetCopySource().

### PR DESCRIPTION
**Pull request by funkyshu**
Ensure that spaces (url encoded or not) in filenames and paths work as expected when copying (particularly for s3 to s3 native copyObject).  Fixes #36

### CODE REVIEW

**Comment by Roosh513** [comment]
Still reading through this. The change seems to deal with way more than space issues and I'm trying to figure out why it was refactored. The side-by-side diff isn't super helpful 🤣

File: backend/s3/file.go [hunk 1]
```diff
@@ -287,48 +303,56 @@ func (f *File) getHeadObject() (*s3.HeadObjectOutput, error) {
...
-		SetBucket(targetFile.bucket).
-		SetCopySource(path.Join(f.bucket, f.key))
-
-	if hasOptions && hasACLOption {
-		copyInput.SetACL(opts.ACL)
-	}
-
 	if isSameAccount {
-		client, err := f.fileSystem.Client()
+		key := strings.Replace(f.key, "%", "%25", -1)
```
[the remark this answers is not in the document]
Comment (jkc) [reply backend/s3/file.go]: he's just encoding the '%' here. '%25' is '%'. '%20' is a space. 

This doesn't answer your questions though...

[the remark this answers is not in the document]
Comment (funkyshu) [reply backend/s3/file.go]: So this only applies to the CopySource value which requires us to URL-encode all values.  Which we do when calling u.EscapedPath().  However when trying to copy the literal "some/file%20withspace.txt", EscapedPath() will not double-encode it.  So what happens is %20 gets interpolated the CopyObjectInput as a space.  Initially I regex'd only literal %20, but then I assumed we'd have the same effect if any other literal url-encoded character was saved to the source filesystem (like what happens to spaces when downloading from the GCS console UI).  
So here I'm assuming that all % signs found are literal and not URL-encoded characters. The backend integration test is what actually found this edge case.
 I'd prefer a consensus on whether this seems right.

[the remark this answers is not in the document]
Comment (funkyshu) [reply backend/s3/file.go]: If it seems ok after my explanation, I can at least add comments to explain my decision in the code.  it's is a bit of "what-the-eff-is-this" code.

[the remark this answers is not in the document]
Comment (dmcilvain44) [reply backend/s3/file.go]: Took me a bit to get my head wrapped around this.  It is a bit odd but I don't see a better way to do it.

[the remark this answers is not in the document]
Comment (Roosh513) [reply backend/s3/file.go]: It looks like `u.EscapedPath()` and `url.PathEscape("foo/bar.txt")` follow different rules on what should be escaped. I tried `url.PathEscape()` and it will double encode your string properly.

https://play.golang.org/p/BnbIhX-JKqR

[the remark this answers is not in the document]
Comment (funkyshu) [reply backend/s3/file.go]: Thanks Chris.  That's a much better solution. My tests still pass.
